torch/csrc/distributed/c10d/NanCheck.cu

```diff
@@ -11,13 +11,134 @@ namespace c10d {
 
 // CUDA kernel to check if data has NAN, device side assert
 // is raised if NAN is found
+
+// Using ulong2 as a "byte pack", with 16 bytes, for efficient data load
+typedef ulong2 BytePack;
+
+// AMD HIP doesn't define `__trap()`, using `assert` instead
+#ifdef USE_ROCM
+#define __trap() assert(0)
+#endif
+
+//// Start of templated functions for checking NaNs inside a BytePack
+
+// (i) General implementation (aka fallback)
+// We use a for loop to iterate over the elements in a BytePack.
+// EltPerPack would be greater than 8 if falling in this case.
+
+template <typename T, int EltPerPack>
+struct CheckBytePack {
```

Comment:
iiuc this generalized kernel would only be used for float8? i guess in a later PR, you would possibly replace this by a specialized one too?

Reply:
Yes